File: src/Difficult/DynamicTest/countDigitOne.java
package Difficult.DynamicTest;

/**
 * @author msc
 * @version 1.0
 * @date 2021/8/13 15:49
 */

/*233. 数字 1 的个数
给定一个整数 n，计算所有小于等于 n 的非负整数中数字 1 出现的个数。



示例 1：

输入：n = 13
输出：6
示例 2：

输入：n = 0
输出：0


提示：

0 <= n <= 2 * 109

* */


public class countDigitOne {

    /*
     *  9->1
     *  99-> 20
     *  999-> 300
     *  9999-> 4000
     *  99999 -> 50000
     *  999999-> 600000
     *   ...
     *  199999999 -> 260000000
     * */
    public int countDigitOne(int n) {
        int[] arr = new int[]{1, 20, 300, 4000, 50000, 600000, 7000000, 80000000, 900000000};
        int temp = n % 10, origin = n;
        int index = 0, multiply = 10, count = 0;
        if (temp >= 1) count++;
        n /= 10;
        while (n != 0) {
            temp = n % 10;
            if (temp == 1) {
                count++;
                count += arr[index];
                count += (origin % multiply);
            } else if (temp > 1) {
                count += multiply;
                count += (arr[index] * temp);
            }
            n /= 10;
            multiply *= 10;
            index++;
        }
        return count;
    }

    /*主要考虑每一个位置上1出现的次数*/
    public int countDigitOne2(int n) {
        int res = 0;
        for (int i = 1; i <= n; i *= 10) {
            int div = i * 10;
            res += (n / div) * i + Math.min(Math.max(n % div - i + 1, 0), i);
        }
        return res;
    }


    public static void main(String[] args) {
        System.out.println(new countDigitOne().countDigitOne(11111));
    }
}

File: src/Difficult/DynamicTest/countDigitOne.py
from typing import List, Tuple, Union
from heapq import heappop, heappush, heapify, heappushpop, heapreplace
from collections import defaultdict, deque, Counter
from itertools import accumulate, permutations, combinations, product, compress
from math import perm, comb, gcd, lcm
from functools import cache, lru_cache, reduce
from sortedcontainers import SortedList, SortedSet, SortedDict
from bisect import bisect_left, bisect_right


class Solution:
    def countDigitOne(self, n: int) -> int:
        s = list(map(int, str(n)))

        @cache
        def f(index, count, isLimit, isNum):
            if index == len(s):
                return count
            res = 0
            if not isNum:
                res += f(index + 1, 0, False, False)
            up = s[index] if isLimit else 9
            for i in range(0 if isNum else 1, up + 1):
                res += f(index + 1, count, isLimit and i == up, True)
            return res

        return f(0, 0, True, False)

    def countDigitOne2(self, n: int) -> int:
        s = str(n)

        @cache
        def dfs(idx, limit, count):
            if idx == len(s):
                return count
            up = int(s[idx]) if limit else 9
            res = 0
            for nxt in range(up + 1):
                res += dfs(idx + 1, limit and nxt == int(s[idx]), count + int(nxt == 1))
            return res

        return dfs(0, True, 0)


if __name__ == '__main__':
    s = Solution()
    print(s.countDigitOne2(2))
